File: Algorithms/Quick Sort/QuickSort.kt
fun quickSort(array: IntArray, left: Int, right: Int) {
    if (array.isEmpty()) return
    if (left >= right) return
    val mid = left + (right - left) / 2
    val el = array[mid]
    var i = left
    var j = right
    while (i <= j) {
        while (array[i] < el) {
            i++
        }
        while (array[j] > el) {
            j--
        }
        if (i <= j) {
            val temp = array[i]
            array[i] = array[j]
            array[j] = temp
            i++
            j--
        }
    }

    if (left < j) {
        quickSort(array, left, j)
    }
    if (right > i) {
        quickSort(array, i, right)
    }
}

fun main() {
    val arr = intArrayOf(123, -3, 12, 567, 0, 12, -1000, 256, 11)
    quickSort(arr, 0, arr.size - 1)
    println(arr.toList())
}

File: Algorithms/Quick Sort/quick_sort.py
def quick_sort(num_list):
   quick_sort_partition(num_list,0,len(num_list)-1)
   return num_list 

   
def quick_sort_partition(num_list,first,last):
   if first<last:
       split_point = partition(num_list,first,last)
       # Left
       quick_sort_partition(num_list,first,split_point-1)
       # Right
       quick_sort_partition(num_list,split_point+1,last)


def partition(num_list,first,last):
   pivot = num_list[first]
   left_mart = first+1
   right_mark = last
   done = False
   while not done:
	   # Find position of pivot
       while left_mart <= right_mark and num_list[left_mart] <= pivot:
           left_mart = left_mart + 1
       while num_list[right_mark] >= pivot and right_mark >= left_mart:
           right_mark = right_mark -1
       if right_mark < left_mart:
           done = True
       else:
           temp = num_list[left_mart]
           num_list[left_mart] = num_list[right_mark]
           num_list[right_mark] = temp
   
   # Pivot position exchange
   temp = num_list[first]
   num_list[first] = num_list[right_mark]
   num_list[right_mark] = temp
   return right_mark


print quick_sort([54,26,93,17,77,31,44,55,20])